File: app/src/main/java/org/linphone/contacts/ContactsManager.java
/*
 * Copyright (c) 2010-2019 Belledonne Communications SARL.
 *
 * This file is part of linphone-android
 * (see https://www.linphone.org).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.linphone.contacts;

import static android.os.AsyncTask.THREAD_POOL_EXECUTOR;

import android.Manifest;
import android.accounts.Account;
import android.accounts.AccountManager;
import android.content.ContentProviderClient;
import android.content.ContentProviderOperation;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.pm.PackageManager;
import android.database.ContentObserver;
import android.database.Cursor;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.os.RemoteException;
import android.provider.ContactsContract;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;
import org.linphone.LinphoneContext;
import org.linphone.LinphoneManager;
import org.linphone.R;
import org.linphone.compatibility.Compatibility;
import org.linphone.core.Address;
import org.linphone.core.Core;
import org.linphone.core.Friend;
import org.linphone.core.FriendList;
import org.linphone.core.FriendListListener;
import org.linphone.core.MagicSearch;
import org.linphone.core.ProxyConfig;
import org.linphone.core.tools.Log;
import org.linphone.settings.LinphonePreferences;

public class ContactsManager extends ContentObserver
        implements FriendListListener, LinphoneContext.CoreStartedListener {
    private List<LinphoneContact> mContacts, mSipContacts;
    private final ArrayList<ContactsUpdatedListener> mContactsUpdatedListeners;
    private MagicSearch mMagicSearch;
    private boolean mContactsFetchedOnce = false;
    private Context mContext;
    private AsyncContactsLoader mLoadContactTask;
    private boolean mInitialized = false;

    public static ContactsManager getInstance() {
        return LinphoneContext.instance().getContactsManager();
    }

    public ContactsManager(Context context) {
        super(new Handler(Looper.getMainLooper()));
        mContext = context;
        mContactsUpdatedListeners = new ArrayList<>();
        mContacts = new ArrayList<>();
        mSipContacts = new ArrayList<>();

        if (LinphoneManager.getCore() != null) {
            mMagicSearch = LinphoneManager.getCore().createMagicSearch();
            mMagicSearch.setLimitedSearch(false); // Do not limit the number of results
        }

        LinphoneContext.instance().addCoreStartedListener(this);
    }

    public void addContactsListener(ContactsUpdatedListener listener) {
        mContactsUpdatedListeners.add(listener);
    }

    public void removeContactsListener(ContactsUpdatedListener listener) {
        mContactsUpdatedListeners.remove(listener);
    }

    public ArrayList<ContactsUpdatedListener> getContactsListeners() {
        return mContactsUpdatedListeners;
    }

    @Override
    public void onChange(boolean selfChange) {
        onChange(selfChange, null);
    }

    @Override
    public void onChange(boolean selfChange, Uri uri) {
        Log.i("[Contacts Manager] Content observer detected a changing in at least one contact");
        fetchContactsAsync();
    }

    @Override
    public void onCoreStarted() {
        // Core has been started, fetch contacts again in case there are some
        // in the configuration file or remote provisioning
        fetchContactsAsync();
    }

    public synchronized List<LinphoneContact> getContacts() {
        return mContacts;
    }

    synchronized void setContacts(List<LinphoneContact> c) {
        mContacts = c;
    }

    public synchronized List<LinphoneContact> getSIPContacts() {
        return mSipContacts;
    }

    synchronized void setSipContacts(List<LinphoneContact> c) {
        mSipContacts = c;
    }

    public void destroy() {
        mContext.getContentResolver().unregisterContentObserver(this);
        LinphoneContext.instance().removeCoreStartedListener(this);

        if (mLoadContactTask != null) {
            mLoadContactTask.cancel(true);
        }
        // LinphoneContact has a Friend field and Friend can have a LinphoneContact has userData
        // Friend also keeps a ref on the Core, so we have to clean them
        for (LinphoneContact c : mContacts) {
            c.setFriend(null);
        }
        mContacts.clear();
        for (LinphoneContact c : mSipContacts) {
            c.setFriend(null);
        }
        mSipContacts.clear();

        Core core = LinphoneManager.getCore();
        if (core != null) {
            for (FriendList list : core.getFriendsLists()) {
                list.removeListener(this);
            }
        }
    }

    public void fetchContactsAsync() {
        if (mLoadContactTask != null) {
            mLoadContactTask.cancel(true);
        }

        if (!hasReadContactsAccess()) {
            Log.w(
                    "[Contacts Manager] Can't fetch native contacts without READ_CONTACTS permission");
        }

        mLoadContactTask = new AsyncContactsLoader(mContext);
        mContactsFetchedOnce = true;
        mLoadContactTask.executeOnExecutor(THREAD_POOL_EXECUTOR);
    }

    public MagicSearch getMagicSearch() {
        return mMagicSearch;
    }

    public boolean contactsFetchedOnce() {
        return mContactsFetchedOnce;
    }

    public List<LinphoneContact> getContacts(String search) {
        search = search.toLowerCase(Locale.getDefault());
        List<LinphoneContact> searchContactsBegin = new ArrayList<>();
        List<LinphoneContact> searchContactsContain = new ArrayList<>();
        for (LinphoneContact contact : getContacts()) {
            if (contact.getFullName() != null) {
                if (contact.getFullName().toLowerCase(Locale.getDefault()).startsWith(search)) {
                    searchContactsBegin.add(contact);
                } else if (contact.getFullName()
                        .toLowerCase(Locale.getDefault())
                        .contains(search)) {
                    searchContactsContain.add(contact);
                }
            }
        }
        searchContactsBegin.addAll(searchContactsContain);
        return searchContactsBegin;
    }

    public List<LinphoneContact> getSIPContacts(String search) {
        search = search.toLowerCase(Locale.getDefault());
        List<LinphoneContact> searchContactsBegin = new ArrayList<>();
        List<LinphoneContact> searchContactsContain = new ArrayList<>();
        for (LinphoneContact contact : getSIPContacts()) {
            if (contact.getFullName() != null) {
                if (contact.getFullName().toLowerCase(Locale.getDefault()).startsWith(search)) {
                    searchContactsBegin.add(contact);
                } else if (contact.getFullName()
                        .toLowerCase(Locale.getDefault())
                        .contains(search)) {
                    searchContactsContain.add(contact);
                }
            }
        }
        searchContactsBegin.addAll(searchContactsContain);
        return searchContactsBegin;
    }

    public void enableContactsAccess() {
        LinphonePreferences.instance().disableFriendsStorage();
    }

    public boolean hasReadContactsAccess() {
        if (mContext == null || mContext.getPackageManager() == null) {
            return false;
        }

        boolean contactsR =
                (PackageManager.PERMISSION_GRANTED
                        == mContext.getPackageManager()
                                .checkPermission(
                                        android.Manifest.permission.READ_CONTACTS,
                                        mContext.getPackageName()));
        return contactsR
                && !mContext.getResources().getBoolean(R.bool.force_use_of_linphone_friends);
    }

    private boolean hasWriteContactsAccess() {
        if (mContext == null) {
            return false;
        }

        return (PackageManager.PERMISSION_GRANTED
                == mContext.getPackageManager()
                        .checkPermission(
                                Manifest.permission.WRITE_CONTACTS, mContext.getPackageName()));
    }

    private boolean hasWriteSyncPermission() {
        if (mContext == null) {
            return false;
        }

        return (PackageManager.PERMISSION_GRANTED
                == mContext.getPackageManager()
                        .checkPermission(
                                Manifest.permission.WRITE_SYNC_SETTINGS,
                                mContext.getPackageName()));
    }

    public boolean isLinphoneContactsPrefered() {
        ProxyConfig lpc = LinphoneManager.getCore().getDefaultProxyConfig();
        return lpc != null
                && lpc.getIdentityAddress()
                        .getDomain()
                        .equals(mContext.getString(R.string.default_domain));
    }

    public void initializeContactManager() {
        if (!mInitialized) {
            if (mContext.getResources().getBoolean(R.bool.use_linphone_tag)) {
                if (hasReadContactsAccess()
                        && hasWriteContactsAccess()
                        && hasWriteSyncPermission()) {
                    if (LinphoneContext.isReady()) {
                        initializeSyncAccount();
                        mInitialized = true;
                    }
                }
            }
        }
    }

    private void makeContactAccountVisible() {
        ContentProviderClient client =
                mContext.getContentResolver()
                        .acquireContentProviderClient(ContactsContract.AUTHORITY_URI);
        ContentValues values = new ContentValues();
        values.put(
                ContactsContract.Settings.ACCOUNT_NAME,
                mContext.getString(R.string.sync_account_name));
        values.put(
                ContactsContract.Settings.ACCOUNT_TYPE,
                mContext.getString(R.string.sync_account_type));
        values.put(ContactsContract.Settings.UNGROUPED_VISIBLE, true);
        try {
            client.insert(
                    ContactsContract.Settings.CONTENT_URI
                            .buildUpon()
                            .appendQueryParameter(ContactsContract.CALLER_IS_SYNCADAPTER, "true")
                            .build(),
                    values);
            Log.i("[Contacts Manager] Contacts account made visible");
        } catch (RemoteException e) {
            Log.e("[Contacts Manager] Couldn't make contacts account visible: " + e);
        }
        Compatibility.closeContentProviderClient(client);
    }

    private void initializeSyncAccount() {
        AccountManager accountManager =
                (AccountManager) mContext.getSystemService(Context.ACCOUNT_SERVICE);

        Account[] accounts =
                accountManager.getAccountsByType(mContext.getString(R.string.sync_account_type));

        if (accounts != null && accounts.length == 0) {
            Account newAccount =
                    new Account(
                            mContext.getString(R.string.sync_account_name),
                            mContext.getString(R.string.sync_account_type));
            try {
                accountManager.addAccountExplicitly(newAccount, null, null);
                Log.i("[Contacts Manager] Contact account added");
                makeContactAccountVisible();
            } catch (Exception e) {
                Log.e("[Contacts Manager] Couldn't initialize sync account: " + e);
            }
        } else if (accounts != null) {
            for (Account account : accounts) {
                Log.i(
                        "[Contacts Manager] Found account with name \""
                                + account.name
                                + "\" and type \""
                                + account.type
                                + "\"");
                makeContactAccountVisible();
            }
        }
    }

    public String getAndroidContactIdFromUri(Uri uri) {
        String[] projection = {ContactsContract.CommonDataKinds.SipAddress.CONTACT_ID};
        Cursor cursor =
                mContext.getApplicationContext()
                        .getContentResolver()
                        .query(uri, projection, null, null, null);
        cursor.moveToFirst();

        int nameColumnIndex =
                cursor.getColumnIndex(ContactsContract.CommonDataKinds.SipAddress.CONTACT_ID);
        String id = cursor.getString(nameColumnIndex);
        cursor.close();
        return id;
    }

    public synchronized LinphoneContact findContactFromAndroidId(String androidId) {
        if (androidId == null) {
            return null;
        }

        for (LinphoneContact c : getContacts()) {
            if (c.getAndroidId() != null && c.getAndroidId().equals(androidId)) {
                return c;
            }
        }
        return null;
    }

    public synchronized LinphoneContact findContactFromAddress(Address address) {
        if (address == null) return null;
        Core core = LinphoneManager.getCore();

        Friend lf = core.findFriend(address);
        if (lf != null) {
            return (LinphoneContact) lf.getUserData();
        }

        String username = address.getUsername();
        if (username == null) {
            Log.w("[Contacts Manager] Address ", address.asString(), " doesn't have a username!");
            return null;
        }

        if (android.util.Patterns.PHONE.matcher(username).matches()) {
            return findContactFromPhoneNumber(username);
        }

        return null;
    }

    public synchronized LinphoneContact findContactFromPhoneNumber(String phoneNumber) {
        if (phoneNumber == null) return null;

        if (!android.util.Patterns.PHONE.matcher(phoneNumber).matches()) {
            Log.w(
                    "[Contacts Manager] Expected phone number but doesn't look like it: "
                            + phoneNumber);
            return null;
        }

        Core core = LinphoneManager.getCore();
        ProxyConfig lpc = null;
        if (core != null) {
            lpc = core.getDefaultProxyConfig();
        }
        if (lpc == null) {
            Log.i("[Contacts Manager] Couldn't find default proxy config...");
            return null;
        }

        String normalized = lpc.normalizePhoneNumber(phoneNumber);
        if (normalized == null) {
            Log.w(
                    "[Contacts Manager] Couldn't normalize phone number "
                            + phoneNumber
                            + ", default proxy config prefix is "
                            + lpc.getDialPrefix());
            normalized = phoneNumber;
        }

        Address addr = lpc.normalizeSipUri(normalized);
        if (addr == null) {
            Log.w("[Contacts Manager] Couldn't normalize SIP URI " + normalized);
            return null;
        }

        // Without this, the hashmap inside liblinphone won't find it...
        addr.setUriParam("user", "phone");
        Friend lf = core.findFriend(addr);
        if (lf != null) {
            return (LinphoneContact) lf.getUserData();
        }

        Log.w("[Contacts Manager] Couldn't find friend...");
        return null;
    }

    public String getAddressOrNumberForAndroidContact(ContentResolver resolver, Uri contactUri) {
        if (resolver == null || contactUri == null) return null;

        // Phone Numbers
        String[] projection = new String[] {ContactsContract.CommonDataKinds.Phone.NUMBER};
        Cursor c = resolver.query(contactUri, projection, null, null, null);
        if (c != null) {
            if (c.moveToNext()) {
                int numberIndex = c.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
                String number = c.getString(numberIndex);
                c.close();
                return number;
            }
            c.close();
        }

        projection = new String[] {ContactsContract.CommonDataKinds.SipAddress.SIP_ADDRESS};
        c = resolver.query(contactUri, projection, null, null, null);
        if (c != null) {
            if (c.moveToNext()) {
                int numberIndex =
                        c.getColumnIndex(ContactsContract.CommonDataKinds.SipAddress.SIP_ADDRESS);
                String address = c.getString(numberIndex);
                c.close();
                return address;
            }
            c.close();
        }

        return null;
    }

    private synchronized boolean refreshSipContact(Friend lf) {
        if (lf == null) return false;
        LinphoneContact contact = (LinphoneContact) lf.getUserData();

        if (contact != null) {
            if (LinphoneContext.instance()
                    .getApplicationContext()
                    .getResources()
                    .getBoolean(R.bool.use_linphone_tag)) {
                // Inserting Linphone information in Android contact if the parameter is enabled
                if (LinphonePreferences.instance()
                        .isPresenceStorageInNativeAndroidContactEnabled()) {
                    // add presence to native contact
                    AsyncContactPresence asyncContactPresence = new AsyncContactPresence(contact);
                    asyncContactPresence.execute();
                }
            }

            if (!mSipContacts.contains(contact)) {
                mSipContacts.add(contact);
                return true;
            }
        }

        return false;
    }

    public void delete(String id) {
        ArrayList<String> ids = new ArrayList<>();
        ids.add(id);
        deleteMultipleContactsAtOnce(ids);
    }

    public void deleteMultipleContactsAtOnce(List<String> ids) {
        String select = ContactsContract.Data.CONTACT_ID + " = ?";
        ArrayList<ContentProviderOperation> ops = new ArrayList<>();

        for (String id : ids) {
            Log.i("[Contacts Manager] Adding Android contact id ", id, " to batch removal");
            String[] args = new String[] {id};
            ops.add(
                    ContentProviderOperation.newDelete(ContactsContract.RawContacts.CONTENT_URI)
                            .withSelection(select, args)
                            .build());
        }

        try {
            mContext.getContentResolver().applyBatch(ContactsContract.AUTHORITY, ops);
        } catch (Exception e) {
            Log.e("[Contacts Manager] " + e);
        }

        // To ensure removed contacts won't appear in the contacts list anymore
        fetchContactsAsync();
    }

    public String getString(int resourceID) {
        if (mContext == null) return null;
        return mContext.getString(resourceID);
    }

    @Override
    public void onContactCreated(FriendList list, Friend lf) {}

    @Override
    public void onContactDeleted(FriendList list, Friend lf) {}

    @Override
    public void onContactUpdated(FriendList list, Friend newFriend, Friend oldFriend) {}

    @Override
    public void onSyncStatusChanged(FriendList list, FriendList.SyncStatus status, String msg) {}

    @Override
    public void onPresenceReceived(FriendList list, Friend[] friends) {
        boolean updated = false;

        for (Friend lf : friends) {
            boolean newContact = refreshSipContact(lf);

            if (newContact) {
                updated = true;
            }
        }

        if (updated) {
            Collections.sort(mSipContacts);
        }

        for (ContactsUpdatedListener listener : mContactsUpdatedListeners) {
            listener.onContactsUpdated();
        }

        Compatibility.createChatShortcuts(mContext);
    }
}

File: app/src/main/java/org/linphone/contact/ContactsManager.kt
/*
 * Copyright (c) 2010-2020 Belledonne Communications SARL.
 *
 * This file is part of linphone-android
 * (see https://www.linphone.org).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.linphone.contact

import android.accounts.Account
import android.accounts.AccountManager
import android.accounts.AuthenticatorDescription
import android.content.ContentResolver
import android.content.ContentUris
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.net.Uri
import android.provider.ContactsContract
import android.util.Patterns
import androidx.core.app.Person
import androidx.core.graphics.drawable.IconCompat
import androidx.lifecycle.MutableLiveData
import java.io.IOException
import kotlinx.coroutines.*
import org.linphone.LinphoneApplication.Companion.coreContext
import org.linphone.LinphoneApplication.Companion.corePreferences
import org.linphone.R
import org.linphone.core.*
import org.linphone.core.tools.Log
import org.linphone.utils.ImageUtils
import org.linphone.utils.PermissionHelper

interface ContactsUpdatedListener {
    fun onContactsUpdated()

    fun onContactUpdated(friend: Friend)
}

open class ContactsUpdatedListenerStub : ContactsUpdatedListener {
    override fun onContactsUpdated() {}

    override fun onContactUpdated(friend: Friend) {}
}

class ContactsManager(private val context: Context) {
    val magicSearch: MagicSearch by lazy {
        val magicSearch = coreContext.core.createMagicSearch()
        magicSearch.limitedSearch = false
        magicSearch
    }

    var latestContactFetch: String = ""

    val fetchInProgress = MutableLiveData<Boolean>()

    var contactIdToWatchFor: String = ""

    val contactAvatar: IconCompat
    val groupAvatar: IconCompat

    private val localFriends = arrayListOf<Friend>()

    private val contactsUpdatedListeners = ArrayList<ContactsUpdatedListener>()

    private val friendListListener: FriendListListenerStub = object : FriendListListenerStub() {
        @Synchronized
        override fun onPresenceReceived(list: FriendList, friends: Array<Friend>) {
            Log.i("[Contacts Manager] Presence received")
            for (friend in friends) {
                refreshContactOnPresenceReceived(friend)
            }
            Log.i("[Contacts Manager] Contacts refreshed due to presence received")
            notifyListeners()
            Log.i("[Contacts Manager] Presence notified to all listeners")
        }
    }

    init {
        initSyncAccount()

        contactAvatar = IconCompat.createWithResource(context, R.drawable.voip_single_contact_avatar_alt)
        groupAvatar = IconCompat.createWithResource(context, R.drawable.voip_multiple_contacts_avatar_alt)

        val core = coreContext.core
        for (list in core.friendsLists) {
            list.addListener(friendListListener)
        }
        Log.i("[Contacts Manager] Created")
    }

    fun shouldDisplaySipContactsList(): Boolean {
        return coreContext.core.defaultAccount?.params?.identityAddress?.domain == corePreferences.defaultDomain
    }

    fun fetchFinished() {
        Log.i("[Contacts Manager] Contacts loader have finished")
        latestContactFetch = System.currentTimeMillis().toString()
        updateLocalContacts()
        fetchInProgress.value = false
        notifyListeners()
    }

    @Synchronized
    fun updateLocalContacts() {
        Log.i("[Contacts Manager] Updating local contact(s)")
        localFriends.clear()

        for (account in coreContext.core.accountList) {
            val friend = coreContext.core.createFriend()
            friend.name = account.params.identityAddress?.displayName ?: account.params.identityAddress?.username

            val address = account.params.identityAddress ?: continue
            friend.address = address

            val pictureUri = corePreferences.defaultAccountAvatarPath
            if (pictureUri != null) {
                val parsedUri = if (pictureUri.startsWith("/")) "file:$pictureUri" else pictureUri
                Log.i("[Contacts Manager] Found local picture URI: $parsedUri")
                friend.photo = parsedUri
            }

            Log.i("[Contacts Manager] Local contact created for account [${address.asString()}] and picture [${friend.photo}]")
            localFriends.add(friend)
        }
    }

    @Synchronized
    fun getAndroidContactIdFromUri(uri: Uri): String? {
        val projection = arrayOf(ContactsContract.Data.CONTACT_ID)
        val cursor = context.contentResolver.query(uri, projection, null, null, null)
        if (cursor?.moveToFirst() == true) {
            val nameColumnIndex = cursor.getColumnIndex(ContactsContract.Data.CONTACT_ID)
            val id = cursor.getString(nameColumnIndex)
            cursor.close()
            return id
        }
        return null
    }

    @Synchronized
    fun findContactById(id: String): Friend? {
        return coreContext.core.defaultFriendList?.findFriendByRefKey(id)
    }

    @Synchronized
    fun findContactByPhoneNumber(number: String): Friend? {
        return coreContext.core.findFriendByPhoneNumber(number)
    }

    @Synchronized
    fun findContactByAddress(address: Address): Friend? {
        for (friend in localFriends) {
            if (friend.address?.weakEqual(address) == true) {
                return friend
            }
        }

        val friend = coreContext.core.findFriend(address)
        if (friend != null) return friend

        val username = address.username
        if (username != null && Patterns.PHONE.matcher(username).matches()) {
            return findContactByPhoneNumber(username)
        }

        return null
    }

    @Synchronized
    fun addListener(listener: ContactsUpdatedListener) {
        contactsUpdatedListeners.add(listener)
    }

    @Synchronized
    fun removeListener(listener: ContactsUpdatedListener) {
        contactsUpdatedListeners.remove(listener)
    }

    @Synchronized
    fun notifyListeners() {
        val list = contactsUpdatedListeners.toMutableList()
        for (listener in list) {
            listener.onContactsUpdated()
        }
    }

    @Synchronized
    fun notifyListeners(friend: Friend) {
        val list = contactsUpdatedListeners.toMutableList()
        for (listener in list) {
            listener.onContactUpdated(friend)
        }
    }

    @Synchronized
    fun destroy() {
        val core = coreContext.core
        for (list in core.friendsLists) list.removeListener(friendListListener)
    }

    private fun initSyncAccount() {
        val accountManager = context.getSystemService(Context.ACCOUNT_SERVICE) as AccountManager
        val accounts = accountManager.getAccountsByType(context.getString(R.string.sync_account_type))
        if (accounts.isEmpty()) {
            val newAccount = Account(
                context.getString(R.string.sync_account_name),
                context.getString(
                    R.string.sync_account_type
                )
            )
            try {
                accountManager.addAccountExplicitly(newAccount, null, null)
                Log.i("[Contacts Manager] Contact account added")
            } catch (e: Exception) {
                Log.e("[Contacts Manager] Couldn't initialize sync account: $e")
            }
        } else {
            for (account in accounts) {
                Log.i("[Contacts Manager] Found account with name [${account.name}] and type [${account.type}]")
            }
        }
    }

    fun getAvailableSyncAccounts(): List<Triple<String, String, Drawable?>> {
        val accountManager = context.getSystemService(Context.ACCOUNT_SERVICE) as AccountManager
        val packageManager = context.packageManager
        val syncAdapters = ContentResolver.getSyncAdapterTypes()
        val authenticators: Array<AuthenticatorDescription> = accountManager.authenticatorTypes
        val available = arrayListOf<Triple<String, String, Drawable?>>()

        for (syncAdapter in syncAdapters) {
            if (syncAdapter.authority == "com.android.contacts" && syncAdapter.isUserVisible) {
                if (syncAdapter.supportsUploading() || syncAdapter.accountType == context.getString(R.string.sync_account_type)) {
                    Log.i("[Contacts Manager] Found sync adapter for com.android.contacts authority: ${syncAdapter.accountType}")
                    val accounts = accountManager.getAccountsByType(syncAdapter.accountType)
                    for (account in accounts) {
                        Log.i("[Contacts Manager] Found account for account type ${syncAdapter.accountType}: ${account.name}")
                        for (authenticator in authenticators) {
                            if (authenticator.type == account.type) {
                                val drawable = packageManager.getDrawable(authenticator.packageName, authenticator.smallIconId, null)
                                val triple = Triple(account.name, account.type, drawable)
                                available.add(triple)
                            }
                        }
                    }
                }
            }
        }

        return available
    }

    @Synchronized
    private fun refreshContactOnPresenceReceived(friend: Friend) {
        Log.d("[Contacts Manager] Received presence information for contact $friend")
        if (corePreferences.storePresenceInNativeContact && PermissionHelper.get().hasWriteContactsPermission()) {
            if (friend.refKey != null) {
                Log.i("[Contacts Manager] Storing presence in native contact ${friend.refKey}")
                storePresenceInNativeContact(friend)
            }
        }
        notifyListeners(friend)
    }

    private fun storePresenceInNativeContact(friend: Friend) {
        val contactEditor = NativeContactEditor(friend)
        for (phoneNumber in friend.phoneNumbers) {
            val sipAddress = friend.getContactForPhoneNumberOrAddress(phoneNumber)
            if (sipAddress != null) {
                Log.d("[Contacts Manager] Found presence information to store in native contact $friend under Linphone sync account")
                contactEditor.setPresenceInformation(
                    phoneNumber,
                    sipAddress
                )
            }
        }
        contactEditor.commit()
    }

    fun createFriendFromSearchResult(searchResult: SearchResult): Friend {
        val searchResultFriend = searchResult.friend
        if (searchResultFriend != null) return searchResultFriend

        val friend = coreContext.core.createFriend()

        val address = searchResult.address
        if (address != null) {
            friend.address = address
        }

        val number = searchResult.phoneNumber
        if (number != null) {
            friend.addPhoneNumber(number)

            if (address != null && address.username == number) {
                friend.removeAddress(address)
            }
        }

        return friend
    }
}

fun Friend.getContactForPhoneNumberOrAddress(value: String): String? {
    val presenceModel = getPresenceModelForUriOrTel(value)
    if (presenceModel != null && presenceModel.basicStatus == PresenceBasicStatus.Open) return presenceModel.contact
    return null
}

fun Friend.hasPresence(): Boolean {
    for (address in addresses) {
        val presenceModel = getPresenceModelForUriOrTel(address.asStringUriOnly())
        if (presenceModel != null && presenceModel.basicStatus == PresenceBasicStatus.Open) return true
    }
    for (number in phoneNumbers) {
        val presenceModel = getPresenceModelForUriOrTel(number)
        if (presenceModel != null && presenceModel.basicStatus == PresenceBasicStatus.Open) return true
    }
    return false
}

fun Friend.getThumbnailUri(): Uri? {
    return getPictureUri(true)
}

fun Friend.getPictureUri(thumbnailPreferred: Boolean = false): Uri? {
    val refKey = refKey
    if (refKey != null) {
        try {
            val lookupUri = ContentUris.withAppendedId(
                ContactsContract.Contacts.CONTENT_URI,
                refKey.toLong()
            )

            if (!thumbnailPreferred) {
                val pictureUri = Uri.withAppendedPath(
                    lookupUri,
                    ContactsContract.Contacts.Photo.DISPLAY_PHOTO
                )
                // Check that the URI points to a real file
                val contentResolver = coreContext.context.contentResolver
                try {
                    if (contentResolver.openAssetFileDescriptor(pictureUri, "r") != null) {
                        return pictureUri
                    }
                } catch (ioe: IOException) { }
            }

            // Fallback to thumbnail if high res picture isn't available
            return Uri.withAppendedPath(
                lookupUri,
                ContactsContract.Contacts.Photo.CONTENT_DIRECTORY
            )
        } catch (e: Exception) { }
    } else if (photo != null) {
        try {
            return Uri.parse(photo)
        } catch (e: Exception) { }
    }
    return null
}

fun Friend.getPerson(): Person {
    val personBuilder = Person.Builder().setName(name)

    val bm: Bitmap? =
        ImageUtils.getRoundBitmapFromUri(
            coreContext.context,
            getThumbnailUri()
        )
    val icon =
        if (bm == null) {
            coreContext.contactsManager.contactAvatar
        } else IconCompat.createWithAdaptiveBitmap(bm)
    if (icon != null) {
        personBuilder.setIcon(icon)
    }

    personBuilder.setUri(nativeUri)
    personBuilder.setImportant(starred)
    return personBuilder.build()
}
